{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "8f7611cb",
      "metadata": {},
      "source": [
        "# TCGA-UCEC — Notebook 05\n",
        "## Tâche supervisée clinique et baseline MLP\n",
        "\n",
        "Objectif : établir une baseline supervisée à l’aide d’un MLP simple\n",
        "en prédisant une variable clinique valide à partir des données RNA-seq."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "id": "97fa23ef",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "✅ Environnement chargé. Working directory: c:\\Z\\M2_AIDA\\TCGA_UCEC_project\n"
          ]
        }
      ],
      "source": [
        "# ==========================================================================================================\n",
        "import os                                       # Navigation fichiers (DIRS, chemins relatifs)\n",
        "import warnings                                 # Masquer warnings (dépréciation)\n",
        "import gc                                       # Gestion mémoire (nettoyage objets inutilisés)\n",
        "import json                                     # Lecture du dictionnaire de métadonnées\n",
        "\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "\n",
        "import pandas as pd\n",
        "\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from sklearn.metrics import roc_auc_score, f1_score, confusion_matrix\n",
        "\n",
        "import tensorflow as tf\n",
        "from tensorflow.keras.models import Sequential\n",
        "from tensorflow.keras.layers import Dense, Dropout\n",
        "from tensorflow.keras.callbacks import EarlyStopping\n",
        "\n",
        "# ==========================================================================================================\n",
        "PROJECT_ROOT = r\"C:\\Z\\M2_AIDA\\TCGA_UCEC_project\" #Laïla\n",
        "\n",
        "DIRS = {\n",
        "    # Racine data\n",
        "    \"DATA\": os.path.join(PROJECT_ROOT, \"data\"),\n",
        "\n",
        "    # Données\n",
        "    \"RAW\": os.path.join(PROJECT_ROOT, \"data\", \"raw\"),\n",
        "\n",
        "    # Données intermédiaires (persistées)\n",
        "    \"PROCESSED\":        os.path.join(PROJECT_ROOT, \"data\", \"processed\"),\n",
        "    \"NORM\":             os.path.join(PROJECT_ROOT, \"data\", \"processed\", \"normalized\"),\n",
        "    \"QC_FILTERED\":      os.path.join(PROJECT_ROOT, \"data\", \"processed\", \"qc_filtered\"),\n",
        "    \"COHORT_FILTERED\":  os.path.join(PROJECT_ROOT, \"data\", \"processed\", \"cohort_filtered\"),\n",
        "    \"SC_OBJECTS\":       os.path.join(PROJECT_ROOT, \"data\", \"processed\", \"single_cell_objects\"),\n",
        "    \"PSEUDOBULK_PROC\":  os.path.join(PROJECT_ROOT, \"data\", \"processed\", \"pseudobulk\"),\n",
        "\n",
        "    # Artefacts (par étape du pipeline)\n",
        "    \"ARTEFACTS\": os.path.join(PROJECT_ROOT, \"data\", \"artefacts\"),\n",
        "\n",
        "    \"EDA\":         os.path.join(PROJECT_ROOT, \"data\", \"artefacts\", \"exploratory_data_analysis\"),\n",
        "    \"QC\":          os.path.join(PROJECT_ROOT, \"data\", \"artefacts\", \"qc_analysis\"),\n",
        "    \"COHORT\":      os.path.join(PROJECT_ROOT, \"data\", \"artefacts\", \"cohort_selection\"),\n",
        "    \"SINGLE_CELL\": os.path.join(PROJECT_ROOT, \"data\", \"artefacts\", \"single_cell_analysis\"),\n",
        "    \"PSEUDOBULK\":  os.path.join(PROJECT_ROOT, \"data\", \"artefacts\", \"pseudobulk_preparation\"),\n",
        "    \"DE\":          os.path.join(PROJECT_ROOT, \"data\", \"artefacts\", \"differential_expression\"),\n",
        "    \"ENRICH\":      os.path.join(PROJECT_ROOT, \"data\", \"artefacts\", \"functional_enrichment\"),\n",
        "\n",
        "    # Autres dossiers du projet\n",
        "    \"RESULTS_R\": os.path.join(PROJECT_ROOT, \"Results_R_Analysis\"),\n",
        "    \"TMP\":       os.path.join(PROJECT_ROOT, \"tmp_cache\"),\n",
        "    \"DOCS\":      os.path.join(PROJECT_ROOT, \"documentation\"),\n",
        "}\n",
        "for path in DIRS.values():\n",
        "    os.makedirs(path, exist_ok=True)\n",
        "os.chdir(PROJECT_ROOT)\n",
        "\n",
        "# ==========================================================================================================\n",
        "warnings.filterwarnings(\"ignore\") \n",
        "plt.rcParams['figure.dpi'] = 100\n",
        "plt.rcParams['savefig.dpi'] = 300\n",
        "plt.rcParams['figure.figsize'] = (10, 6) \n",
        "\n",
        "# ==========================================================================================================\n",
        "print(f\"✅ Environnement chargé. Working directory: {os.getcwd()}\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4e2bdce8",
      "metadata": {},
      "source": [
        "## 1. Chargement des données patient-level\n",
        "\n",
        "Les données chargées ici correspondent aux sorties du Notebook 04"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "id": "81a8da99",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "✅ Donnees chargees\n",
            "   - Expression : (31876, 557) (genes x patients)\n",
            "   - Clinique   : (597, 78) (patients x variables)\n"
          ]
        }
      ],
      "source": [
        "expr_filename = \"expr_patient_level_tcga_ucec.tsv\"\n",
        "clin_filename = \"clin_patient_level_tcga_ucec.tsv\"\n",
        "\n",
        "expr_path = os.path.join(DIRS[\"PROCESSED\"], expr_filename)\n",
        "clin_path = os.path.join(DIRS[\"PROCESSED\"], clin_filename)\n",
        "\n",
        "expr = pd.read_csv(expr_path, sep=\"\\t\", index_col=0)\n",
        "clin = pd.read_csv(clin_path, sep=\"\\t\", index_col=0)\n",
        "\n",
        "print(\n",
        "    f\"✅ Donnees chargees\\n\"\n",
        "    f\"   - Expression : {expr.shape} (genes x patients)\\n\"\n",
        "    f\"   - Clinique   : {clin.shape} (patients x variables)\"\n",
        ")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "id": "e12116ae",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Nombre de colonnes dupliquées (expr): 0\n",
            "Exemples de patients dupliqués (expr):\n",
            "Index([], dtype='object')\n"
          ]
        }
      ],
      "source": [
        "dup_expr = expr.columns[expr.columns.duplicated()]\n",
        "print(\"Nombre de colonnes dupliquées (expr):\", len(dup_expr))\n",
        "print(\"Exemples de patients dupliqués (expr):\")\n",
        "print(dup_expr[:10])\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 14,
      "id": "b7250c3f",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Distribution du nombre de samples par patient (expr):\n",
            "count\n",
            "1    557\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "expr_counts = expr.columns.value_counts()\n",
        "\n",
        "print(\"Distribution du nombre de samples par patient (expr):\")\n",
        "print(expr_counts.value_counts().sort_index())\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "69d40352",
      "metadata": {},
      "source": [
        "## 2. Définition de la tâche supervisée\n",
        "\n",
        "Les sous-types moléculaires TCGA (POLE / MSI / CN-low / CN-high)\n",
        "ne sont pas directement disponibles dans les données cliniques Xena\n",
        "pour la cohorte UCEC.\n",
        "\n",
        "Nous définissons donc une tâche supervisée alternative, biologiquement\n",
        "pertinente et reproductible : **la prédiction du grade tumoral**\n",
        "(`tumor_grade.diagnoses`) à partir des profils d'expression RNA-seq.\n",
        "\n",
        "Ce choix est courant dans les analyses transcriptomiques exploratoires\n",
        "et permet d'évaluer la capacité du modèle à capturer un signal clinique\n",
        "à partir des données moléculaires."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "id": "8d250999",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "dict_keys(['__name__', '__doc__', '__package__', '__loader__', '__spec__', '__builtin__', '__builtins__', '_ih', '_oh', '_dh', 'In', 'Out', 'get_ipython', 'exit', 'quit', 'open', '_', '__', '___', '__vsc_ipynb_file__', '_i', '_ii', '_iii', '_i1', 'os', 'warnings', 'gc', 'json', 'np', 'plt', 'pd', 'train_test_split', 'LabelEncoder', 'roc_auc_score', 'f1_score', 'confusion_matrix', 'tf', 'Sequential', 'Dense', 'Dropout', 'EarlyStopping', 'PROJECT_ROOT', 'DIRS', 'path', '_i2', '_i3', 'expr_filename', 'clin_filename', 'expr_path', 'clin_path', 'expr', 'clin', '_i4', '_i5', '_i6', 'expr_id_lengths', '_i7', '_i8', '_i9', 'i', '_i10', 'dup_expr', '_i11', 'expr_counts', '_i12', '_i13', '_i14', '_i15', '_i16', '_i17', '_i18', '_i19', '_i20', 'dup_expr_patients', '_i21', '_i22'])\n"
          ]
        }
      ],
      "source": [
        "print(globals().keys())\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "id": "0061e9e9",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "tumor_grade.diagnoses in columns: True\n",
            "\n",
            "Valeurs brutes :\n",
            "TCGA-2E-A9G8    Not Reported\n",
            "TCGA-4E-A92E    Not Reported\n",
            "TCGA-5B-A90C    Not Reported\n",
            "TCGA-5S-A9Q8    Not Reported\n",
            "TCGA-A5-A0G1    Not Reported\n",
            "TCGA-A5-A0G2    Not Reported\n",
            "TCGA-A5-A0G3    Not Reported\n",
            "TCGA-A5-A0G5    Not Reported\n",
            "TCGA-A5-A0G9    Not Reported\n",
            "TCGA-A5-A0GA    Not Reported\n",
            "TCGA-A5-A0GB    Not Reported\n",
            "TCGA-A5-A0GD    Not Reported\n",
            "TCGA-A5-A0GE    Not Reported\n",
            "TCGA-A5-A0GG    Not Reported\n",
            "TCGA-A5-A0GH    Not Reported\n",
            "TCGA-A5-A0GI    Not Reported\n",
            "TCGA-A5-A0GJ    Not Reported\n",
            "TCGA-A5-A0GM    Not Reported\n",
            "TCGA-A5-A0GN    Not Reported\n",
            "TCGA-A5-A0GP    Not Reported\n",
            "Name: tumor_grade.diagnoses, dtype: object\n",
            "\n",
            "Distribution :\n",
            "tumor_grade.diagnoses\n",
            "Not Reported    585\n",
            "NaN              12\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "print(\"tumor_grade.diagnoses in columns:\",\n",
        "      \"tumor_grade.diagnoses\" in clin.columns)\n",
        "\n",
        "print(\"\\nValeurs brutes :\")\n",
        "print(clin[\"tumor_grade.diagnoses\"].head(20))\n",
        "\n",
        "print(\"\\nDistribution :\")\n",
        "print(clin[\"tumor_grade.diagnoses\"].value_counts(dropna=False))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 25,
      "id": "a392093b",
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "sample                                                       597\n",
              "sample_id.samples                                            597\n",
              "pathology_report_uuid.samples                                558\n",
              "case_id                                                      557\n",
              "id                                                           557\n",
              "updated_datetime.treatments.diagnoses                        545\n",
              "created_datetime.treatments.diagnoses                        545\n",
              "submitter_id.treatments.diagnoses                            545\n",
              "treatment_id.treatments.diagnoses                            545\n",
              "age_at_diagnosis.diagnoses                                   533\n",
              "age_at_earliest_diagnosis_in_years.diagnoses.xena_derived    533\n",
              "age_at_earliest_diagnosis.diagnoses.xena_derived             533\n",
              "days_to_birth.demographic                                    533\n",
              "days_to_last_follow_up.diagnoses                             453\n",
              "days_to_collection.samples                                   448\n",
              "initial_weight.samples                                        96\n",
              "days_to_death.demographic                                     90\n",
              "year_of_birth.demographic                                     60\n",
              "age_at_index.demographic                                      59\n",
              "case_submitter_id.annotations                                 33\n",
              "dtype: int64"
            ]
          },
          "execution_count": 25,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "clin.nunique().sort_values(ascending=False).head(20)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 26,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Nb fichiers candidats: 4\n",
            "C:\\Z\\M2_AIDA\\TCGA_UCEC_project\\data\\processed\\clin_patient_level_tcga_ucec.tsv\n",
            "C:\\Z\\M2_AIDA\\TCGA_UCEC_project\\data\\processed\\clin_tcga_ucec.tsv\n",
            "C:\\Z\\M2_AIDA\\TCGA_UCEC_project\\data\\processed\\expr_norm_tcga_ucec.tsv\n",
            "C:\\Z\\M2_AIDA\\TCGA_UCEC_project\\data\\processed\\expr_patient_level_tcga_ucec.tsv\n"
          ]
        }
      ],
      "source": [
        "from pathlib import Path\n",
        "import re\n",
        "\n",
        "# adapte si ton project root est ailleurs\n",
        "ROOT = Path(\".\").resolve()\n",
        "\n",
        "patterns = [\n",
        "    re.compile(r\"subtype|subtypes|molecular|po le|pole|msi|cn[-_ ]?low|cn[-_ ]?high|copy|serous|endometr\", re.I),\n",
        "    re.compile(r\"ucec|uter\", re.I),\n",
        "]\n",
        "\n",
        "candidates = []\n",
        "for p in ROOT.rglob(\"*\"):\n",
        "    if p.is_file() and p.suffix.lower() in [\".tsv\", \".csv\", \".txt\", \".xlsx\", \".json\"]:\n",
        "        name = p.name\n",
        "        if patterns[0].search(name) or patterns[1].search(name):\n",
        "            candidates.append(p)\n",
        "\n",
        "print(\"Nb fichiers candidats:\", len(candidates))\n",
        "for p in candidates[:50]:\n",
        "    print(p)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "id": "f394cbdb",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Index(['Unnamed: 0', 'sample', 'id', 'disease_type', 'case_id', 'primary_site',\n",
            "       'alcohol_history.exposures', 'race.demographic', 'gender.demographic',\n",
            "       'ethnicity.demographic', 'vital_status.demographic',\n",
            "       'age_at_index.demographic', 'days_to_birth.demographic',\n",
            "       'year_of_birth.demographic', 'year_of_death.demographic',\n",
            "       'primary_site.project', 'project_id.project', 'disease_type.project',\n",
            "       'name.project', 'name.program.project',\n",
            "       'tissue_source_site_id.tissue_source_site', 'code.tissue_source_site',\n",
            "       'name.tissue_source_site', 'project.tissue_source_site',\n",
            "       'bcr_id.tissue_source_site', 'days_to_death.demographic',\n",
            "       'entity_submitter_id.annotations', 'notes.annotations',\n",
            "       'submitter_id.annotations', 'classification.annotations',\n",
            "       'entity_id.annotations', 'created_datetime.annotations',\n",
            "       'annotation_id.annotations', 'entity_type.annotations',\n",
            "       'updated_datetime.annotations', 'case_id.annotations',\n",
            "       'state.annotations', 'category.annotations', 'status.annotations',\n",
            "       'case_submitter_id.annotations', 'figo_stage.diagnoses',\n",
            "       'synchronous_malignancy.diagnoses', 'days_to_diagnosis.diagnoses',\n",
            "       'last_known_disease_status.diagnoses',\n",
            "       'tissue_or_organ_of_origin.diagnoses',\n",
            "       'days_to_last_follow_up.diagnoses', 'age_at_diagnosis.diagnoses',\n",
            "       'primary_diagnosis.diagnoses', 'prior_malignancy.diagnoses',\n",
            "       'year_of_diagnosis.diagnoses', 'prior_treatment.diagnoses',\n",
            "       'morphology.diagnoses', 'classification_of_tumor.diagnoses',\n",
            "       'icd_10_code.diagnoses', 'site_of_resection_or_biopsy.diagnoses',\n",
            "       'tumor_grade.diagnoses', 'progression_or_recurrence.diagnoses',\n",
            "       'age_at_earliest_diagnosis.diagnoses.xena_derived',\n",
            "       'age_at_earliest_diagnosis_in_years.diagnoses.xena_derived',\n",
            "       'updated_datetime.treatments.diagnoses',\n",
            "       'treatment_id.treatments.diagnoses',\n",
            "       'submitter_id.treatments.diagnoses',\n",
            "       'treatment_type.treatments.diagnoses', 'state.treatments.diagnoses',\n",
            "       'treatment_or_therapy.treatments.diagnoses',\n",
            "       'created_datetime.treatments.diagnoses', 'sample_type_id.samples',\n",
            "       'tumor_descriptor.samples', 'sample_id.samples', 'sample_type.samples',\n",
            "       'composition.samples', 'days_to_collection.samples',\n",
            "       'initial_weight.samples', 'preservation_method.samples',\n",
            "       'pathology_report_uuid.samples', 'oct_embedded.samples',\n",
            "       'specimen_type.samples', 'is_ffpe.samples', 'tissue_type.samples'],\n",
            "      dtype='object')\n"
          ]
        }
      ],
      "source": [
        "df = pd.read_csv(\n",
        "    r\"C:\\Z\\M2_AIDA\\TCGA_UCEC_project\\data\\processed\\clin_patient_level_tcga_ucec.tsv\",\n",
        "    sep=\"\\t\"\n",
        ")\n",
        "\n",
        "print(df.columns)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 28,
      "id": "91a44169",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "figo_stage.diagnoses in columns: True\n",
            "\n",
            "Valeurs brutes (head):\n",
            "TCGA-2E-A9G8      Stage III\n",
            "TCGA-4E-A92E        Stage I\n",
            "TCGA-5B-A90C       Stage IA\n",
            "TCGA-5S-A9Q8     Stage IIIA\n",
            "TCGA-A5-A0G1       Stage IA\n",
            "TCGA-A5-A0G2     Stage IIIB\n",
            "TCGA-A5-A0G3    Stage IIIC2\n",
            "TCGA-A5-A0G5       Stage IB\n",
            "TCGA-A5-A0G9       Stage IB\n",
            "TCGA-A5-A0GA    Stage IIIC2\n",
            "TCGA-A5-A0GB       Stage IB\n",
            "TCGA-A5-A0GD       Stage IA\n",
            "TCGA-A5-A0GE       Stage IA\n",
            "TCGA-A5-A0GG       Stage IA\n",
            "TCGA-A5-A0GH       Stage IA\n",
            "TCGA-A5-A0GI       Stage IA\n",
            "TCGA-A5-A0GJ       Stage IA\n",
            "TCGA-A5-A0GM       Stage IA\n",
            "TCGA-A5-A0GN       Stage IB\n",
            "TCGA-A5-A0GP       Stage IA\n",
            "Name: figo_stage.diagnoses, dtype: object\n",
            "\n",
            "Distribution:\n",
            "figo_stage.diagnoses\n",
            "Stage IA       171\n",
            "Stage IB       160\n",
            "Stage IIIA      43\n",
            "Stage IIIC      39\n",
            "Stage II        35\n",
            "Stage IC        27\n",
            "Stage IIIC1     24\n",
            "Stage IVB       24\n",
            "Stage IIIC2     22\n",
            "Stage IIB       14\n",
            "NaN             12\n",
            "Stage IIIB       7\n",
            "Stage IIA        6\n",
            "Stage IV         4\n",
            "Stage I          3\n",
            "Stage IVA        3\n",
            "Stage III        3\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "print(\"figo_stage.diagnoses in columns:\",\n",
        "      \"figo_stage.diagnoses\" in clin.columns)\n",
        "\n",
        "print(\"\\nValeurs brutes (head):\")\n",
        "print(clin[\"figo_stage.diagnoses\"].head(20))\n",
        "\n",
        "print(\"\\nDistribution:\")\n",
        "print(clin[\"figo_stage.diagnoses\"].value_counts(dropna=False))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "id": "737cda4e",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "figo_stage.diagnoses\n",
            "Early    585\n",
            "None      12\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "# Mapping Early / Late (sans encore entraîner quoi que ce soit)\n",
        "def map_stage(stage):\n",
        "    if pd.isna(stage):\n",
        "        return None\n",
        "    if stage.startswith(\"Stage I\") or stage.startswith(\"Stage II\"):\n",
        "        return \"Early\"\n",
        "    if stage.startswith(\"Stage III\") or stage.startswith(\"Stage IV\"):\n",
        "        return \"Late\"\n",
        "    return None\n",
        "\n",
        "y_stage = clin[\"figo_stage.diagnoses\"].apply(map_stage)\n",
        "\n",
        "print(y_stage.value_counts(dropna=False))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 32,
      "id": "82af40de",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "figo_stage.diagnoses\n",
            "Early    554\n",
            "Late      31\n",
            "None      12\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "def map_stage(stage):\n",
        "    if pd.isna(stage):\n",
        "        return None\n",
        "    stage = stage.strip()\n",
        "    if stage.startswith(\"Stage I \") or stage in [\"Stage I\", \"Stage IA\", \"Stage IB\", \"Stage IC\"]:\n",
        "        return \"Early\"\n",
        "    if stage.startswith(\"Stage II\"):\n",
        "        return \"Early\"\n",
        "    if stage.startswith(\"Stage III\") or stage.startswith(\"Stage IV\"):\n",
        "        return \"Late\"\n",
        "    return None\n",
        "\n",
        "y_stage = clin[\"figo_stage.diagnoses\"].apply(map_stage)\n",
        "print(y_stage.value_counts(dropna=False))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 34,
      "id": "a8a6f2eb",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "figo_stage.diagnoses\n",
            "Early    585\n",
            "None      12\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "def map_stage_v2(stage):\n",
        "    if pd.isna(stage):\n",
        "        return None\n",
        "    stage = stage.strip()\n",
        "    if stage.startswith(\"Stage I\"):\n",
        "        return \"Early\"\n",
        "    if stage.startswith(\"Stage III\") or stage.startswith(\"Stage IV\"):\n",
        "        return \"Late\"\n",
        "    return None  # exclut Stage II\n",
        "\n",
        "y_stage_v2 = clin[\"figo_stage.diagnoses\"].apply(map_stage_v2)\n",
        "print(y_stage_v2.value_counts(dropna=False))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 35,
      "id": "e1ac4df2",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "figo_stage.diagnoses\n",
            "Early    585\n",
            "None      12\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "import re\n",
        "\n",
        "def map_stage_clean(stage):\n",
        "    if pd.isna(stage):\n",
        "        return None\n",
        "    stage = stage.strip()\n",
        "\n",
        "    # extraire le chiffre romain principal\n",
        "    m = re.match(r\"Stage\\s+(I|II|III|IV)\", stage)\n",
        "    if not m:\n",
        "        return None\n",
        "\n",
        "    main_stage = m.group(1)\n",
        "\n",
        "    if main_stage == \"I\":\n",
        "        return \"Early\"\n",
        "    if main_stage in [\"III\", \"IV\"]:\n",
        "        return \"Late\"\n",
        "    return None  # exclut Stage II\n",
        "\n",
        "y_stage = clin[\"figo_stage.diagnoses\"].apply(map_stage_clean)\n",
        "print(y_stage.value_counts(dropna=False))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 36,
      "id": "5edf5da7",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "figo_stage.diagnoses\n",
            "None     587\n",
            "Late       7\n",
            "Early      3\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "import re\n",
        "\n",
        "def map_stage_clean(stage):\n",
        "    if pd.isna(stage):\n",
        "        return None\n",
        "    stage = stage.strip()\n",
        "\n",
        "    # ORDRE IMPORTANT : III / IV avant I\n",
        "    m = re.match(r\"Stage\\s+(III|IV|II|I)\\b\", stage)\n",
        "    if not m:\n",
        "        return None\n",
        "\n",
        "    main_stage = m.group(1)\n",
        "\n",
        "    if main_stage == \"I\":\n",
        "        return \"Early\"\n",
        "    if main_stage in [\"III\", \"IV\"]:\n",
        "        return \"Late\"\n",
        "    return None  # exclut Stage II\n",
        "\n",
        "y_stage = clin[\"figo_stage.diagnoses\"].apply(map_stage_clean)\n",
        "print(y_stage.value_counts(dropna=False))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 39,
      "id": "3b16ef7c",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "figo_stage.diagnoses\n",
            "Stage IA       171\n",
            "Stage IB       160\n",
            "Stage IIIA      43\n",
            "Stage IIIC      39\n",
            "Stage II        35\n",
            "Stage IC        27\n",
            "Stage IIIC1     24\n",
            "Stage IVB       24\n",
            "Stage IIIC2     22\n",
            "Stage IIB       14\n",
            "NaN             12\n",
            "Stage IIIB       7\n",
            "Stage IIA        6\n",
            "Stage IV         4\n",
            "Stage I          3\n",
            "Stage IVA        3\n",
            "Stage III        3\n",
            "Name: count, dtype: int64\n"
          ]
        }
      ],
      "source": [
        "print(clin[\"figo_stage.diagnoses\"].value_counts(dropna=False))\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "66679085",
      "metadata": {},
      "source": [
        "## Définition de la tâche supervisée (FIGO binaire)\n",
        "\n",
        "Les sous-types moléculaires TCGA n’étant pas directement disponibles dans les\n",
        "données cliniques Xena pour la cohorte UCEC, une tâche supervisée alternative\n",
        "a été définie.\n",
        "\n",
        "Nous utilisons le **stade FIGO**, binarisé en :\n",
        "- **Early stage** (I–II),\n",
        "- **Late stage** (III–IV).\n",
        "\n",
        "Ce choix est biologiquement pertinent en carcinome de l’endomètre, largement\n",
        "utilisé en pratique clinique, et permet d’évaluer la capacité des profils\n",
        "RNA-seq à prédire un phénotype clinique majeur lié à la progression tumorale.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 41,
      "id": "844469db",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Patients finaux : 585\n",
            "Labels finaux   : 585\n",
            "figo_stage.diagnoses\n",
            "Early    416\n",
            "Late     169\n",
            "Name: count, dtype: int64\n",
            "X shape : (585, 31876)\n",
            "y shape : (585,)\n"
          ]
        }
      ],
      "source": [
        "# Repartir des patients communs expr / clin\n",
        "common_patients = expr.columns.intersection(clin.index)\n",
        "\n",
        "expr_aligned = expr.loc[:, common_patients]\n",
        "clin_aligned = clin.loc[common_patients]\n",
        "\n",
        "# Mapping FIGO binaire SUR clin_aligned uniquement\n",
        "labels = clin_aligned[label_col].map(\n",
        "    lambda x: \"Early\" if x in early_stages else\n",
        "              \"Late\" if x in late_stages else np.nan\n",
        ")\n",
        "\n",
        "# Suppression des patients sans label\n",
        "labels = labels.dropna()\n",
        "\n",
        "# ALIGNEMENT FINAL (CRITIQUE)\n",
        "expr_final = expr_aligned.loc[:, labels.index]\n",
        "\n",
        "# Vérifications\n",
        "print(\"Patients finaux :\", expr_final.shape[1])\n",
        "print(\"Labels finaux   :\", labels.shape[0])\n",
        "print(labels.value_counts())\n",
        "\n",
        "# Construction X / y\n",
        "X = expr_final.T.values\n",
        "\n",
        "le = LabelEncoder()\n",
        "y = le.fit_transform(labels)\n",
        "\n",
        "print(\"X shape :\", X.shape)\n",
        "print(\"y shape :\", y.shape)\n",
        "\n",
        "assert X.shape[0] == y.shape[0]\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c31e0d5b",
      "metadata": {},
      "source": [
        "## 3. Encodage des labels"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 42,
      "id": "1a306dbb",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Classes encodees : ['Early', 'Late']\n",
            "X shape : (585, 31876)\n",
            "y shape : (585,)\n"
          ]
        }
      ],
      "source": [
        "# Encodage des labels (APRÈS alignement final)\n",
        "\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "\n",
        "le = LabelEncoder()\n",
        "y = le.fit_transform(labels)\n",
        "\n",
        "# Construction de X à partir de expr_final UNIQUEMENT\n",
        "X = expr_final.T.values  # patients x genes\n",
        "\n",
        "print(\"Classes encodees :\", list(le.classes_))\n",
        "print(\"X shape :\", X.shape)\n",
        "print(\"y shape :\", y.shape)\n",
        "\n",
        "assert X.shape[0] == y.shape[0]\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "27c80877",
      "metadata": {},
      "source": [
        "## 4. Split train / test"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "3d95cf36",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Train shape : (468, 31876) (468,)\n",
            "Test shape  : (117, 31876) (117,)\n",
            "Train label distribution : [333 135]\n",
            "Test label distribution  : [83 34]\n"
          ]
        }
      ],
      "source": [
        "# =====================================================================================\n",
        "# 4. Split train / test (APRÈS alignement et encodage validés)\n",
        "\n",
        "from sklearn.model_selection import train_test_split\n",
        "\n",
        "# fixe le générateur aléatoire pour garantir la reproductibilité :\n",
        "# à chaque exécution, les tirages aléatoires donnent exactement les mêmes résultats.\n",
        "X_train, X_test, y_train, y_test = train_test_split(\n",
        "    X,\n",
        "    y,\n",
        "    test_size=0.2,\n",
        "    stratify=y,\n",
        "    random_state=42\n",
        ")\n",
        "\n",
        "print(\"Train shape :\", X_train.shape, y_train.shape)\n",
        "print(\"Test shape  :\", X_test.shape, y_test.shape)\n",
        "\n",
        "# Vérification distribution des classes\n",
        "print(\"Train label distribution :\", np.bincount(y_train))\n",
        "print(\"Test label distribution  :\", np.bincount(y_test))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 44,
      "id": "a8e74f6c",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "expr cols: 557\n",
            "labels n: 585\n",
            "expr unique cols: 557\n",
            "labels unique idx: 545\n"
          ]
        }
      ],
      "source": [
        "print(\"expr cols:\", expr.shape[1])\n",
        "print(\"labels n:\", labels.shape[0])\n",
        "print(\"expr unique cols:\", expr.columns.nunique())\n",
        "print(\"labels unique idx:\", labels.index.nunique())\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c9093ef4",
      "metadata": {},
      "source": [
        "## 5. Baseline supervisée — MLP minimal"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 45,
      "id": "5d7ff1ca",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Epoch 1/100\n",
            "12/12 [==============================] - 2s 84ms/step - loss: 6.0519 - accuracy: 0.6417 - val_loss: 6.1618 - val_accuracy: 0.5745\n",
            "Epoch 2/100\n",
            "12/12 [==============================] - 1s 62ms/step - loss: 3.7463 - accuracy: 0.7701 - val_loss: 4.9873 - val_accuracy: 0.7234\n",
            "Epoch 3/100\n",
            "12/12 [==============================] - 1s 71ms/step - loss: 2.1702 - accuracy: 0.8743 - val_loss: 5.2502 - val_accuracy: 0.6915\n",
            "Epoch 4/100\n",
            "12/12 [==============================] - 1s 59ms/step - loss: 1.5091 - accuracy: 0.9198 - val_loss: 12.2354 - val_accuracy: 0.6170\n",
            "Epoch 5/100\n",
            "12/12 [==============================] - 1s 59ms/step - loss: 2.9145 - accuracy: 0.9037 - val_loss: 5.4911 - val_accuracy: 0.7234\n",
            "Epoch 6/100\n",
            "12/12 [==============================] - 1s 49ms/step - loss: 1.5850 - accuracy: 0.9037 - val_loss: 4.6450 - val_accuracy: 0.7340\n",
            "Epoch 7/100\n",
            "12/12 [==============================] - 1s 58ms/step - loss: 0.9700 - accuracy: 0.9385 - val_loss: 6.0373 - val_accuracy: 0.7447\n",
            "Epoch 8/100\n",
            "12/12 [==============================] - 1s 64ms/step - loss: 0.6970 - accuracy: 0.9439 - val_loss: 7.2927 - val_accuracy: 0.6809\n",
            "Epoch 9/100\n",
            "12/12 [==============================] - 1s 63ms/step - loss: 0.7524 - accuracy: 0.9385 - val_loss: 5.6046 - val_accuracy: 0.7234\n",
            "Epoch 10/100\n",
            "12/12 [==============================] - 1s 65ms/step - loss: 0.6837 - accuracy: 0.9599 - val_loss: 6.8523 - val_accuracy: 0.7021\n",
            "Epoch 11/100\n",
            "12/12 [==============================] - 1s 59ms/step - loss: 0.2806 - accuracy: 0.9679 - val_loss: 8.1206 - val_accuracy: 0.7128\n",
            "Epoch 12/100\n",
            "12/12 [==============================] - 1s 62ms/step - loss: 0.3223 - accuracy: 0.9679 - val_loss: 7.2344 - val_accuracy: 0.6702\n",
            "Epoch 13/100\n",
            "12/12 [==============================] - 1s 60ms/step - loss: 0.1680 - accuracy: 0.9813 - val_loss: 8.1068 - val_accuracy: 0.6702\n",
            "Epoch 14/100\n",
            "12/12 [==============================] - 1s 62ms/step - loss: 0.6048 - accuracy: 0.9492 - val_loss: 5.3721 - val_accuracy: 0.7447\n",
            "Epoch 15/100\n",
            "12/12 [==============================] - 1s 63ms/step - loss: 0.2953 - accuracy: 0.9759 - val_loss: 11.1203 - val_accuracy: 0.6596\n",
            "Epoch 16/100\n",
            "12/12 [==============================] - 1s 60ms/step - loss: 0.2438 - accuracy: 0.9733 - val_loss: 6.8064 - val_accuracy: 0.7340\n"
          ]
        }
      ],
      "source": [
        "model = Sequential([\n",
        "    Dense(64, activation=\"relu\", input_shape=(X_train.shape[1],)),\n",
        "    Dropout(0.3),\n",
        "    Dense(len(le.classes_), activation=\"softmax\")\n",
        "])\n",
        "\n",
        "model.compile(\n",
        "    optimizer=\"adam\",\n",
        "    loss=\"sparse_categorical_crossentropy\",\n",
        "    metrics=[\"accuracy\"]\n",
        ")\n",
        "\n",
        "early_stop = EarlyStopping(\n",
        "    monitor=\"val_loss\",\n",
        "    patience=10,\n",
        "    restore_best_weights=True\n",
        ")\n",
        "\n",
        "history = model.fit(\n",
        "    X_train, y_train,\n",
        "    validation_split=0.2,\n",
        "    epochs=100,\n",
        "    batch_size=32,\n",
        "    callbacks=[early_stop],\n",
        "    verbose=1\n",
        ")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "f0c8c872",
      "metadata": {},
      "source": [
        "Le MLP apprend très rapidement à séparer les classes sur le jeu d’entraînement,\n",
        "mais généralise difficilement, ce qui indique un sur-apprentissage lié à la très\n",
        "haute dimension des données transcriptomiques et au déséquilibre des classes."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "73386c54",
      "metadata": {},
      "source": [
        "## 6. Évaluation du modèle"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 46,
      "id": "69fb5d1a",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "4/4 [==============================] - 0s 33ms/step\n",
            "ROC-AUC       : 0.763\n",
            "F1-score macro: 0.719\n",
            "Matrice de confusion :\n",
            "[[67 16]\n",
            " [12 22]]\n"
          ]
        }
      ],
      "source": [
        "from sklearn.metrics import f1_score, confusion_matrix, roc_auc_score\n",
        "\n",
        "# Probabilités pour ROC-AUC\n",
        "y_proba = model.predict(X_test)[:, 1]\n",
        "y_pred = (y_proba >= 0.5).astype(int)\n",
        "\n",
        "f1 = f1_score(y_test, y_pred, average=\"macro\")\n",
        "auc = roc_auc_score(y_test, y_proba)\n",
        "cm = confusion_matrix(y_test, y_pred)\n",
        "\n",
        "print(f\"ROC-AUC       : {auc:.3f}\")\n",
        "print(f\"F1-score macro: {f1:.3f}\")\n",
        "print(\"Matrice de confusion :\")\n",
        "print(cm)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "1e88c81d",
      "metadata": {},
      "source": [
        "**Évaluation du MLP supervisé (FIGO binaire)**\n",
        "\n",
        "Le modèle MLP a été évalué sur un jeu de test indépendant (20 % des patients), en utilisant des métriques adaptées à une classification binaire déséquilibrée (Early vs Late).\n",
        "\n",
        "Performances obtenues sur le jeu de test :\n",
        "\n",
        "ROC-AUC = 0.763, indiquant une capacité de discrimination significativement supérieure au hasard.\n",
        "\n",
        "F1-score macro = 0.719, reflétant un compromis satisfaisant entre précision et rappel pour les deux classes.\n",
        "\n",
        "Matrice de confusion :\n",
        "\n",
        "La majorité des patients Early est correctement classée.\n",
        "\n",
        "Les patients Late sont détectés avec une sensibilité modérée, ce qui est attendu compte tenu du déséquilibre des classes.\n",
        "\n",
        "Ces résultats montrent que le MLP parvient à extraire un signal clinique pertinent à partir des profils transcriptomiques RNA-seq, malgré la très haute dimension des données (≈ 31 000 gènes) et un effectif limité.\n",
        "Toutefois, une comparaison avec une baseline plus simple (régression logistique) est nécessaire afin de déterminer si la complexité du réseau de neurones est réellement justifiée."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "32966740",
      "metadata": {},
      "source": [
        "## 7. Baseline supervisée — Régression logistique\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 47,
      "id": "90697088",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Baseline Logistic Regression — ROC-AUC       : 0.811\n",
            "Baseline Logistic Regression — F1-score macro: 0.731\n",
            "Matrice de confusion (baseline) :\n",
            "[[56 27]\n",
            " [ 3 31]]\n"
          ]
        }
      ],
      "source": [
        "from sklearn.linear_model import LogisticRegression\n",
        "from sklearn.metrics import roc_auc_score, f1_score, confusion_matrix\n",
        "\n",
        "# Modèle baseline : régression logistique\n",
        "logreg = LogisticRegression(\n",
        "    penalty=\"l2\",\n",
        "    solver=\"liblinear\",\n",
        "    max_iter=1000,\n",
        "    class_weight=\"balanced\",\n",
        "    random_state=42\n",
        ")\n",
        "\n",
        "# Entraînement\n",
        "logreg.fit(X_train, y_train)\n",
        "\n",
        "# Prédictions\n",
        "y_proba_lr = logreg.predict_proba(X_test)[:, 1]\n",
        "y_pred_lr = (y_proba_lr >= 0.5).astype(int)\n",
        "\n",
        "# Évaluation\n",
        "auc_lr = roc_auc_score(y_test, y_proba_lr)\n",
        "f1_lr = f1_score(y_test, y_pred_lr, average=\"macro\")\n",
        "cm_lr = confusion_matrix(y_test, y_pred_lr)\n",
        "\n",
        "print(f\"Baseline Logistic Regression — ROC-AUC       : {auc_lr:.3f}\")\n",
        "print(f\"Baseline Logistic Regression — F1-score macro: {f1_lr:.3f}\")\n",
        "print(\"Matrice de confusion (baseline) :\")\n",
        "print(cm_lr)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "1d9a9933",
      "metadata": {},
      "source": [
        "**Comparaison des modèles supervisés : MLP vs baseline linéaire**\n",
        "\n",
        "Une régression logistique a été entraînée comme baseline supervisée sur le même jeu de données, en utilisant le même split train/test et les mêmes métriques d’évaluation que pour le MLP.\n",
        "\n",
        "Performances sur le jeu de test :\n",
        "Régression logistique (baseline)\n",
        "ROC-AUC = 0.811\n",
        "F1-score macro = 0.731\n",
        "MLP\n",
        "ROC-AUC = 0.763\n",
        "F1-score macro = 0.719\n",
        "\n",
        "La baseline linéaire obtient de meilleures performances que le réseau de neurones, tant en termes de capacité de discrimination (ROC-AUC) que d’équilibre entre précision et rappel (F1 macro)."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "1232cb3c",
      "metadata": {},
      "source": [
        "**Interprétation**\n",
        "\n",
        "Ces résultats indiquent que, dans ce cadre, le signal transcriptomique associé au stade FIGO (Early vs Late) est majoritairement linéairement séparable.\n",
        "L’ajout de complexité via un MLP n’apporte pas de gain de performance et conduit même à un léger sur-apprentissage, probablement en raison :\n",
        "de la très haute dimension des données RNA-seq (~31 000 gènes),\n",
        "du nombre limité d’échantillons,\n",
        "et du déséquilibre entre les classes.\n",
        "\n",
        "✅ Conclusion méthodologique\n",
        "La régression logistique est retenue comme modèle de référence, car elle offre :\n",
        "de meilleures performances,\n",
        "une plus grande stabilité,\n",
        "et une interprétabilité directe des coefficients, facilitant l’analyse biologique.\n",
        "Le MLP est conservé comme point de comparaison, mais sa complexité n’est pas justifiée dans ce contexte."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "c14832db",
      "metadata": {},
      "source": [
        "## 8. Interprétation biologique — Régression logistique\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "04647abf",
      "metadata": {},
      "source": [
        "### Extraire les coefficients du modèle\n",
        "\n",
        "Donne les 20 gènes les plus discriminants."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 48,
      "id": "03e2a9cf",
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene</th>\n",
              "      <th>coef</th>\n",
              "      <th>abs_coef</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>30528</th>\n",
              "      <td>ENSG00000283052.1</td>\n",
              "      <td>0.039369</td>\n",
              "      <td>0.039369</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24566</th>\n",
              "      <td>ENSG00000253582.1</td>\n",
              "      <td>0.036462</td>\n",
              "      <td>0.036462</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28098</th>\n",
              "      <td>ENSG00000270775.1</td>\n",
              "      <td>-0.036312</td>\n",
              "      <td>0.036312</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27604</th>\n",
              "      <td>ENSG00000267986.1</td>\n",
              "      <td>-0.036176</td>\n",
              "      <td>0.036176</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21900</th>\n",
              "      <td>ENSG00000236048.2</td>\n",
              "      <td>0.035773</td>\n",
              "      <td>0.035773</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31087</th>\n",
              "      <td>ENSG00000286329.1</td>\n",
              "      <td>0.035054</td>\n",
              "      <td>0.035054</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31118</th>\n",
              "      <td>ENSG00000286391.1</td>\n",
              "      <td>-0.034934</td>\n",
              "      <td>0.034934</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25390</th>\n",
              "      <td>ENSG00000257539.2</td>\n",
              "      <td>-0.034905</td>\n",
              "      <td>0.034905</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19623</th>\n",
              "      <td>ENSG00000227403.2</td>\n",
              "      <td>-0.034752</td>\n",
              "      <td>0.034752</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20267</th>\n",
              "      <td>ENSG00000229914.1</td>\n",
              "      <td>-0.034662</td>\n",
              "      <td>0.034662</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19583</th>\n",
              "      <td>ENSG00000227242.4</td>\n",
              "      <td>-0.034518</td>\n",
              "      <td>0.034518</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21960</th>\n",
              "      <td>ENSG00000236285.1</td>\n",
              "      <td>0.034193</td>\n",
              "      <td>0.034193</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20837</th>\n",
              "      <td>ENSG00000232133.1</td>\n",
              "      <td>0.033422</td>\n",
              "      <td>0.033422</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21270</th>\n",
              "      <td>ENSG00000233754.2</td>\n",
              "      <td>-0.033402</td>\n",
              "      <td>0.033402</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25743</th>\n",
              "      <td>ENSG00000259050.2</td>\n",
              "      <td>-0.032857</td>\n",
              "      <td>0.032857</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24571</th>\n",
              "      <td>ENSG00000253614.1</td>\n",
              "      <td>-0.032699</td>\n",
              "      <td>0.032699</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20387</th>\n",
              "      <td>ENSG00000230406.2</td>\n",
              "      <td>-0.032587</td>\n",
              "      <td>0.032587</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25570</th>\n",
              "      <td>ENSG00000258524.1</td>\n",
              "      <td>0.032553</td>\n",
              "      <td>0.032553</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25858</th>\n",
              "      <td>ENSG00000259392.2</td>\n",
              "      <td>-0.032385</td>\n",
              "      <td>0.032385</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21599</th>\n",
              "      <td>ENSG00000234952.2</td>\n",
              "      <td>0.032373</td>\n",
              "      <td>0.032373</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    gene      coef  abs_coef\n",
              "30528  ENSG00000283052.1  0.039369  0.039369\n",
              "24566  ENSG00000253582.1  0.036462  0.036462\n",
              "28098  ENSG00000270775.1 -0.036312  0.036312\n",
              "27604  ENSG00000267986.1 -0.036176  0.036176\n",
              "21900  ENSG00000236048.2  0.035773  0.035773\n",
              "31087  ENSG00000286329.1  0.035054  0.035054\n",
              "31118  ENSG00000286391.1 -0.034934  0.034934\n",
              "25390  ENSG00000257539.2 -0.034905  0.034905\n",
              "19623  ENSG00000227403.2 -0.034752  0.034752\n",
              "20267  ENSG00000229914.1 -0.034662  0.034662\n",
              "19583  ENSG00000227242.4 -0.034518  0.034518\n",
              "21960  ENSG00000236285.1  0.034193  0.034193\n",
              "20837  ENSG00000232133.1  0.033422  0.033422\n",
              "21270  ENSG00000233754.2 -0.033402  0.033402\n",
              "25743  ENSG00000259050.2 -0.032857  0.032857\n",
              "24571  ENSG00000253614.1 -0.032699  0.032699\n",
              "20387  ENSG00000230406.2 -0.032587  0.032587\n",
              "25570  ENSG00000258524.1  0.032553  0.032553\n",
              "25858  ENSG00000259392.2 -0.032385  0.032385\n",
              "21599  ENSG00000234952.2  0.032373  0.032373"
            ]
          },
          "execution_count": 48,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "# noms des gènes = index de expr_final\n",
        "gene_names = expr_final.index\n",
        "\n",
        "# coefficients du modèle logistique\n",
        "coefs = logreg.coef_[0]\n",
        "\n",
        "coef_df = pd.DataFrame({\n",
        "    \"gene\": gene_names,\n",
        "    \"coef\": coefs\n",
        "})\n",
        "\n",
        "# tri par importance absolue\n",
        "coef_df[\"abs_coef\"] = coef_df[\"coef\"].abs()\n",
        "coef_df = coef_df.sort_values(\"abs_coef\", ascending=False)\n",
        "\n",
        "coef_df.head(20)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "20f40537",
      "metadata": {},
      "source": [
        "### Séparer Early vs Late\n",
        "\n",
        "coef < 0 → associé à Early\n",
        "coef > 0 → associé à Late"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 49,
      "id": "70c76f26",
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gènes associés Early FIGO :\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene</th>\n",
              "      <th>coef</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>28098</th>\n",
              "      <td>ENSG00000270775.1</td>\n",
              "      <td>-0.036312</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27604</th>\n",
              "      <td>ENSG00000267986.1</td>\n",
              "      <td>-0.036176</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31118</th>\n",
              "      <td>ENSG00000286391.1</td>\n",
              "      <td>-0.034934</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25390</th>\n",
              "      <td>ENSG00000257539.2</td>\n",
              "      <td>-0.034905</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19623</th>\n",
              "      <td>ENSG00000227403.2</td>\n",
              "      <td>-0.034752</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20267</th>\n",
              "      <td>ENSG00000229914.1</td>\n",
              "      <td>-0.034662</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19583</th>\n",
              "      <td>ENSG00000227242.4</td>\n",
              "      <td>-0.034518</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21270</th>\n",
              "      <td>ENSG00000233754.2</td>\n",
              "      <td>-0.033402</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25743</th>\n",
              "      <td>ENSG00000259050.2</td>\n",
              "      <td>-0.032857</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24571</th>\n",
              "      <td>ENSG00000253614.1</td>\n",
              "      <td>-0.032699</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    gene      coef\n",
              "28098  ENSG00000270775.1 -0.036312\n",
              "27604  ENSG00000267986.1 -0.036176\n",
              "31118  ENSG00000286391.1 -0.034934\n",
              "25390  ENSG00000257539.2 -0.034905\n",
              "19623  ENSG00000227403.2 -0.034752\n",
              "20267  ENSG00000229914.1 -0.034662\n",
              "19583  ENSG00000227242.4 -0.034518\n",
              "21270  ENSG00000233754.2 -0.033402\n",
              "25743  ENSG00000259050.2 -0.032857\n",
              "24571  ENSG00000253614.1 -0.032699"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Gènes associés Late FIGO :\n"
          ]
        },
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene</th>\n",
              "      <th>coef</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>30528</th>\n",
              "      <td>ENSG00000283052.1</td>\n",
              "      <td>0.039369</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24566</th>\n",
              "      <td>ENSG00000253582.1</td>\n",
              "      <td>0.036462</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21900</th>\n",
              "      <td>ENSG00000236048.2</td>\n",
              "      <td>0.035773</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31087</th>\n",
              "      <td>ENSG00000286329.1</td>\n",
              "      <td>0.035054</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21960</th>\n",
              "      <td>ENSG00000236285.1</td>\n",
              "      <td>0.034193</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20837</th>\n",
              "      <td>ENSG00000232133.1</td>\n",
              "      <td>0.033422</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25570</th>\n",
              "      <td>ENSG00000258524.1</td>\n",
              "      <td>0.032553</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21599</th>\n",
              "      <td>ENSG00000234952.2</td>\n",
              "      <td>0.032373</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13420</th>\n",
              "      <td>ENSG00000176857.5</td>\n",
              "      <td>0.032036</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19448</th>\n",
              "      <td>ENSG00000226756.1</td>\n",
              "      <td>0.031686</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    gene      coef\n",
              "30528  ENSG00000283052.1  0.039369\n",
              "24566  ENSG00000253582.1  0.036462\n",
              "21900  ENSG00000236048.2  0.035773\n",
              "31087  ENSG00000286329.1  0.035054\n",
              "21960  ENSG00000236285.1  0.034193\n",
              "20837  ENSG00000232133.1  0.033422\n",
              "25570  ENSG00000258524.1  0.032553\n",
              "21599  ENSG00000234952.2  0.032373\n",
              "13420  ENSG00000176857.5  0.032036\n",
              "19448  ENSG00000226756.1  0.031686"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        }
      ],
      "source": [
        "top_early = coef_df[coef_df[\"coef\"] < 0].head(10)\n",
        "top_late  = coef_df[coef_df[\"coef\"] > 0].head(10)\n",
        "\n",
        "print(\"Gènes associés Early FIGO :\")\n",
        "display(top_early[[\"gene\", \"coef\"]])\n",
        "\n",
        "print(\"Gènes associés Late FIGO :\")\n",
        "display(top_late[[\"gene\", \"coef\"]])\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "0c97b387",
      "metadata": {},
      "source": [
        "### Interprétation biologique des gènes discriminants\n",
        "\n",
        "L’analyse des coefficients de la régression logistique met en évidence un ensemble de gènes dont l’expression est associée au stade FIGO (Early vs Late). Les gènes les plus contributifs identifiés par le modèle correspondent majoritairement à des transcrits faiblement ou non annotés (identifiants ENSG), suggérant que la discrimination entre stades repose sur des signatures transcriptomiques globales plutôt que sur quelques gènes canoniques isolés.\n",
        "\n",
        "Les gènes associés aux stades Late (III–IV) présentent des coefficients positifs, indiquant une surexpression relative dans les tumeurs plus avancées. Bien que l’annotation fonctionnelle directe de ces transcrits soit limitée, leur contribution collective est compatible avec l’activation de programmes biologiques complexes impliqués dans la progression tumorale, tels que l’augmentation de l’activité transcriptionnelle, la plasticité cellulaire et l’instabilité génomique, caractéristiques des carcinomes endométriaux avancés.\n",
        "\n",
        "À l’inverse, les gènes associés aux stades Early (I–II) montrent des coefficients négatifs, traduisant une expression plus élevée dans les tumeurs de bas stade. Cette signature est cohérente avec des profils transcriptomiques correspondant à des tissus plus différenciés et à des états tumoraux moins agressifs.\n",
        "\n",
        "Dans l’ensemble, ces résultats indiquent que le stade FIGO est associé à des différences d’expression à large échelle, détectables par un modèle linéaire simple. L’absence de gènes fortement annotés parmi les plus discriminants souligne les limites de l’interprétation gène-par-gène et met en évidence l’intérêt d’une approche transcriptomique globale pour caractériser la progression tumorale dans le carcinome de l’endomètre.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "9701ca0b",
      "metadata": {},
      "source": [
        "## 9. Prochaines étapes\n",
        "\n",
        "- Évaluer la **robustesse des performances** par runs répétés ou validation croisée.\n",
        "- Explorer un **tuning léger des hyperparamètres** du MLP afin de vérifier si un gain robuste est possible.\n",
        "- Approfondir l’**interprétation biologique** à partir des coefficients de la régression logistique retenue comme modèle de référence.\n",
        "- Mettre en œuvre un **modèle non supervisé (autoencodeur)** pour explorer la structure latente des profils transcriptomiques et leur lien avec les stades FIGO.\n"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python (tcga_tf)",
      "language": "python",
      "name": "tcga_tf"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.8"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
